### ISSUE TITLE
Offer a multi-instance Redis+Sentinel plan

### ISSUE THREAD

**Issue by sharms** [18F/cg-product#101]
In order to ensure apps using our Redis service aren't disrupted when we updates k8s, we will provide a multi-instance Redis+Sentinel plan that uses k8s affinity controls to distribute pods across nodes.

**Comment by cnelson** [comment]
Noting for when we prioritize this card, that not having HA redis, caused a dashboard outage: https://cloudgov.statuspage.io/incidents/mtnwp3vcngvw

**Comment by rogeruiz** [comment]
Taking a look at these examples later today.https://github.com/fabric8io/kansible/tree/master/vendor/k8s.io/kubernetes/examples/redis

Reading through https://redis.io/topics/sentinel

**Comment by rogeruiz** [comment]
My previous examples of k8s-redis-sentinel were out of date. but I think I found some that I can use https://github.com/rogeruiz/k8s-redis and https://github.com/rogeruiz/k8s-redis-1

**Comment by rogeruiz** [comment]
I've been running into issues with the scheduler and available resources on k8s Development. It's prevented me from successfully debugging the deployment of a `3x-ha` redis plan via the broker.

**Comment by mogul** [comment]
Do we need to split the actual sizing of this plan out into a separate card?

**Comment by rogeruiz** [comment]
Nah, I don't think so @mogul but we do need to figure out what the sizings for the plans are going to be. right now we have 3 servers and 3 sentinels (minimum for sentinels, best practice for servers) and 2 proxies offered here. Does the HA Redis plan actually need more than that? I mean I guess I could see the number of servers going up, but not sure if it really matters much to bump up everything else.

I always felt like this story was more about availability and not changing the plans. I'm not sure what small, medium, large instances of Redis would offer folks. 💭

**Comment by cnelson** [comment]
Given that our current redis plan is super-tiny right right now I don't think we need to worry about doing more sizing on the servers than that we are currently offering [(512mb max ram, with infinite oversubcribe)](https://github.com/18F/kubernetes-broker/blob/master/catalogData/redis28/standard/k8s/deployment.json#L38-L44) although we might want to guarantee each server some amount of ram.

Have we profiled the sentinel and proxy nodes to understand their resource needs? We should set them to the smallest amount they require while making sure the entire cluster of 8 containers will  stack neatly onto our minions.

Currently our minions have 30004M of ram allocated to pods. So for example (numbers except for server are made up, I have no idea what sentinel/proxy require):
```
3x Server @ 512M 
3x Sentinel @ 64M 
2x Proxy @ 64M
```
would let us neatly stack 16 instances of this plan on each minion.

**Comment by cnelson** [comment]
Spent a few hours playing with this and found several issues:

The proxy appears to have a memory leak. 
---
Every connection opened causes the proxy to use more memory that isn't released when the connection is closed. I put ~ 20,000 connections through the proxy and it was OOMKilled

This needs to be resolved before we can go live with this.

The proxy doesn't use timeouts for it's TCP connections.
---
More than once I saw my redis client connected to a dead socket that wouldn't disconnect me, and wouldn't respond to commands either.  This looks like it's happening when the master has died, but the proxy hasn't noticed yet so it's making a connection without a timeout to an IP/port that doesn't respond anymore.

WDYT about using DialTimeout() instead of DialTCP() with a very short timeout (say 50ms)?  If we can't connection to a pod on the same network in a few milliseconds something is very wrong

Pods are limiting memory but not requesting it
---

WDYT about requesting memory so we don't oversubscribe nodes too much?

Anti-affinity doesn't appear to be to working. 
---
For example I saw three sentinels  and both proxies all scheduled on the same node
> xef91aa21601e4-sentinel-289727132-5v2t8    1/1       Running   0          1m        x.x.x.14   ip-y-y-y--232.us-gov-west-1.compute.internal
> xef91aa21601e4-sentinel-289727132-6b02h    1/1       Running   0          37m       x.x.x.6    ip-y-y-y--232.us-gov-west-1.compute.internal
> xef91aa21601e4-sentinel-289727132-92wx7    1/1       Running   0          1m        x.x.x.18   ip-y-y-y--232.us-gov-west-1.compute.internal

> xef91aa21601e4-proxy-483865637-z3ndg       1/1       Running   1          25m       x.x.x.11   ip-y-y-y-232.us-gov-west-1.compute.internal
> xef91aa21601e4-proxy-483865637-zdsz5       1/1       Running   0          6m        x.x.x.19   ip-y-y-y-232.us-gov-west-1.compute.internal


If k8s is slow starting the masters, the whole instance ends up being slow to start
---

The proxies and sentinels start first and quickly end up 
 CrashLoopBackup state because the crash repeatedly until a master is up.
	
WDYT about implementing some retry logic in the sentinel and proxy startup scripts so they don't die immediately if the masters aren't up?

**Comment by rogeruiz** [comment]
Awesome. thanks @cnelson. could you walk me through the tests you did on this sometime this week? most of these _seem_ simple enough to fix quickly. 👍

**Comment by cnelson** [comment]
Let's pair on this in the AM but in case you want to reproduce before that, the tests were pretty simple:

I used `create-service`, while it was provisioning the instance I was running `kubectl get pods -o wide | grep :id:` to watch it come up.  That's how I noticed the `CrashLoopBackoff`

Then I `bind-service` to a random app. Then ssh'd into the app and:

## To create tons of connections
`while [ 1 ] ; do nc -z ${REDIS} ${REDIS_PORT}; let C=C+1; if (( $C % 10000 == 0 )); then echo $C; fi done`

While that was running I was just running top in the proxy containers while running top via `kubectl exec` to watch the resource utilization

## to create useful connections

`while /bin/true; do echo -e "AUTH ...\r\nGET x" | nc ${REDIS} ${REDIS_PORT}; done`

While this was running I was killing one pod at a time and waiting for it to recover, or recreating an entire minion via bosh.

This is where I noticed the Dial() timeout problem (at least I think that's what it is).  

After a few iterations of killing pods, I noticed the anti-affinity issue.

**Comment by rogeruiz** [comment]
Updated list of PRs that need merging on Monday.

feature patches:
- https://github.com/18F/cg-deploy-kubernetes/pull/91 (acceptance tests and orchestration)
- https://github.com/18F/kubernetes-broker/pull/57 (k8s manifest and Docker images)
- https://github.com/18F/cg-site/pull/1015 (docs)
- https://github.com/18F/redis-sentinel-proxy/pull/3 (proxy updates, binary is built as `1.0.1`)
